### PR TITLE
TzKT reward & block provider

### PR DESCRIPTION
**Pull request by m-kus**
This is an implementation of the `BlockApi` and `RewardApi` interfaces for [TzKT API](https://api.tzkt.io/).  
All methods and dry-run are covered with tests, although not suited for running in travis pipeline (takes about 10min), probably one can use dumps from Rpc/Tzstats providers instead of querying them each time.
  
Please note, that there can be differences in final payout amounts due to how different providers handle different rewards:  
* Tzstats provider does not handle accusation rewards/losses
* Both RPC and Tzstats calculates incorrect staking balances for cycles with snapshot@16 (see https://medium.com/@_MisterWalker_/we-all-were-wrong-baking-bad-and-most-bakers-were-using-wrong-data-to-calculate-staking-rewards-a8c26f5ec62b)
  
Also, TzKT allows to distinct between cases when rewards are missed due to overdelegation and other reasons. Also, you can track the order of delegations and determine which particular delegator is not to be paid. This can probably help with #69 and #204 
  
Generally, it would be great if TRD supports the [Bakers Registry](https://hackmd.io/DZyJU5HSThmGX8ER7tprXg?both) config format in order to unlock the whole potential of TzKT/Tzstats reward splits.  

Would love to answer any questions or add more tests if needed :)

### CODE REVIEW

**Comment by m-kus** [comment]
For some reason Travis build machine cannot reach our API host :( Anyways, I've disabled those tests on travis, still they can be run locally.  
Also added more tests for edge cases and file cache for `rpc` / `tzstats` providers output, now it takes only few seconds to run.  
  
One more thing, I've added an extra cmdline argument for passing custom API endpoint (for local instances of TzKT/Tzstats), it's None by default so should not break anything.  

Please review and let me know if one need to fix/add/remove anything.

**Comment by jdsika** [comment]
@amzid can you help with Travis ci?
Do you need some kind of whitelisting to reach the server?

**Comment by amzid** [comment]
> @amzid can you help with Travis ci?
> Do you need some kind of whitelisting to reach the server?

I am also not sure why this happens. I would suggest contacting the support of Travis CI and sending the log of the unsuccessful build of commit 9f7a3e2.

**Comment by Groxan** [comment]
Any updates on this PR?

**Comment by amzid** [comment]
For me the tests are working locally as well. I contacted the travis CI support, if we do not receive a response within the next week, we might need to deactivate the tests on the CI.

**Comment by amzid** [comment]
The Travis CI support suggests trying the following work around https://stackoverflow.com/a/33184416 and is ready to follow up on this issue.

**Comment by m-kus** [comment]
I surely can mock TzKT requests, but it would make less sense since it doesn't guarantee TzKT won't respond differently in the future. I think disabling theese tests running on Travis is a preferable option, let me know which of two options works best for you.

**Comment by amzid** [comment]
In this case, I think the implementation overhead is not worth it since it is not guaranteed to be stable. Since the tests can be anyway performed locally, I personally also opt for the option of disabling them in Travis.

File: src/tzkt/tzkt_block_api.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+from tzkt.tzkt_api import TzKTApi, TzKTApiError
+from api.block_api import BlockApi
+
+
+class TzKTBlockApiImpl(BlockApi):
+
+    def __init__(self, nw, verbose=False, base_url=None):
+        super(TzKTBlockApiImpl, self).__init__(nw)
+        if base_url is None:
+            self.api = TzKTApi.from_network(nw['NAME'].lower(), verbose=verbose)
```
Comment (amzid) [inline src/tzkt/tzkt_block_api.py:10]: In the current implementation of the TRD, we fix the development network name to Alphanet, but it is meant to be the current active test network. I think tzkt expects the name of the active test network and not alphanet, right?

**Review by amzid — APPROVED** [review]
This PR is ready to be merged from my side. In future PRs, we will work on a strategy to isolate tests from the core code.
Was this work already paid in the context of a grant/company income, or should the work effort be compensated by the TRD grant?

**Comment by m-kus** [comment]
> This PR is ready to be merged from my side. In future PRs, we will work on a strategy to isolate tests from the core code.
> Was this work already paid in the context of a grant/company income, or should the work effort be compensated by the TRD grant?

Great! No additional compensation is needed.

**Comment by jdsika** [comment]
Are the tests now running locally with a commit hook?
